File: examples/Portfolio_LP.cpp
#include "helix/lp_solver.hpp"

#include "portfolio_common.hpp"

#include <Eigen/Dense>

#include <exception>
#include <iostream>

int main() {
    const helix_examples::PortfolioInput input = helix_examples::make_example_input();

    helix::SolverSettings settings;
    settings.absolute_tolerance = 1e-7;
    settings.relative_tolerance = 1e-7;
    settings.polish = true;
    helix::LpSolver solver(settings);

    try {
        const helix_examples::PortfolioConstraints constraints =
            helix_examples::make_portfolio_constraints(input);

        helix::LpProblem problem;
        problem.c = Eigen::VectorXd::Zero(2 * input.alpha.size());
        problem.c.head(input.alpha.size()) = -input.alpha;
        problem.A = constraints.matrix;
        problem.lower = constraints.lower;
        problem.upper = constraints.upper;

        helix_examples::PortfolioResult result =
            helix_examples::decode_result(input, solver.solve(problem));
        if (!result.solve_result.success()) {
            std::cerr << "portfolio LP failed: " << result.solve_result.message << '\n';
            return 1;
        }
        helix_examples::print_result(input, result);
    } catch (const std::exception& error) {
        std::cerr << "invalid portfolio input: " << error.what() << '\n';
        return 2;
    }
    return 0;
}

File: examples/python/portfolio_lp.py
import numpy as np

import helix


settings = helix.SolverSettings()
settings.absolute_tolerance = 1e-7
settings.relative_tolerance = 1e-7

optimizer = helix.PortfolioLpOptimizer(settings)
result = optimizer.solve(
    alpha=np.array([0.02, 0.08, 0.05, -0.01]),
    current_position_value=np.array([300_000.0, 250_000.0, 250_000.0, 200_000.0]),
    max_position_weight=np.array([0.35, 0.40, 0.30, 0.25]),
    liquidity_limit_value=np.array([100_000.0, 120_000.0, 80_000.0, 60_000.0]),
    max_turnover_ratio=0.10,
    capital=1_000_000.0,
)

if not result.success:
    raise RuntimeError(result.solver_result.message)

print("target weights:", result.target_weight)
print("trade values:", result.trade_value)
print("turnover ratio:", result.turnover_ratio)
